napari_clusters_plotter/_plotter.py

```diff
@@ -795,3 +791,137 @@ def run(
+            or self.visualized_layer.name not in self.viewer.layers
+        ):
+            self.viewer.add_layer(self.visualized_layer)
+        else:
+            layer_in_viewer = self.viewer.layers[self.visualized_layer.name]
+            layer_in_viewer.data = self.visualized_layer.data
+            if isinstance(self.visualized_layer, Points):
+                layer_in_viewer.face_color = self.visualized_layer.face_color
+            elif isinstance(self.visualized_layer, Surface):
+                layer_in_viewer.colormap = self.visualized_layer.colormap
+            elif isinstance(self.visualized_layer, Labels):
+                layer_in_viewer.color = self.visualized_layer.color
+            else:
+                print("Update failed")
+
+    def _draw_cluster_image(
+        self,
+        is_tracking_data: bool,
+        plot_cluster_name: str,
+        cluster_ids,
+        cmap_dict=None,
+    ) -> Layer:
+        from matplotlib.colors import to_rgba_array
+
+        from ._utilities import (
+            generate_cluster_4d_labels,
+            generate_cluster_image,
+            generate_cluster_surface,
+            generate_cluster_tracks,
+            get_nice_colormap,
+            get_surface_color_map,
+        )
+
+        """
+        Generate the cluster image layer.
+        """
+        nice_colormap = get_nice_colormap()
+        napari_colormap = get_surface_color_map(max(cluster_ids))
+
+        if (
+            isinstance(self.analysed_layer, Labels)
+            and len(self.analysed_layer.data.shape) == 4
+            and not is_tracking_data
+        ):
+            cluster_data = generate_cluster_4d_labels(
+                self.analysed_layer, plot_cluster_name
+            )
+
+            cluster_layer = Layer.create(
+                cluster_data,
+                {
+                    "color": cmap_dict,
+                    "name": "cluster_ids_in_space",
+                    "scale": self.layer_select.value.scale,
+                },
+            )
+
+        elif (
+            isinstance(self.analysed_layer, Labels)
+            and len(self.analysed_layer.data.shape) == 4
+            and is_tracking_data
+        ):
+            cluster_data = generate_cluster_tracks(
+                self.analysed_layer, plot_cluster_name
+            )
+
+            cluster_layer = Layer.create(
+                cluster_data,
+                {
+                    "color": cmap_dict,
+                    "name": "cluster_ids_in_space",
+                    "scale": self.layer_select.value.scale,
+                },
+            )
+
+        elif isinstance(self.analysed_layer, Surface):
+            cluster_data = generate_cluster_surface(
+                self.analysed_layer.data, self.cluster_ids
+            )
+
+            cluster_layer = Layer.create(
+                cluster_data,
+                {
+                    "contrast_limits": [0, self.cluster_ids.max() + 1],
```

Comment:
@jo-mueller 
I think the weird surface data glitch I found might be caused by the line for the contrast limits. Everytime you add a new color it would change the limits and possibly influence all colors. I don't know enough about the surface layers yet though to know why this line might be needed.....

Reply:
@Cryaaa Ok, that's weird. I tried to reproduce it with the example data from above and it looks perfectly fine for me 🤔

Reply:
@jo-mueller,
damn that is strange.... I'll have to investigate my environment then and see what might be influencing it but maybe should not try to change things too drastically

Reply:
Same - I'll create a fresh environment to see whether it still works alright.

Reply:
In another environment I have the same problem. Maybe to add some more details, I was adding new clusters each time in the video (i.e. hold shift while circling the same points each time). If I just circle the same points without holding shift it behaves as you would expect

Reply:
@Cryaaa ok, now I actually see what the bug is, by adding a color you mean adding a new class while holding the shift key? I'm not sure whether I would treat that as a bug rather than a feature. After all, adding a new color is supposed to add a new class color - selecting the same set of points with the shift-key pressed would only a bit of an...exotic? way of using that feature ^^

Maybe we could make the resetting the colormaps a bit more prominent with an extra button somewhere in the plot? Currently this is implicitly done by clicking somewhere in the image or making an empty selection without holding the shift-key.

Reply:
@jo-mueller,
let's talk this through in person later, then I can show you how the bug might be problematic!

Reply:
@Cryaaa could you maybe share the code snipped you used to create the problematic sample data you showed earlier?

Reply:
Sure

```python
import napari
import vedo
import pandas as pd
import napari_process_points_and_surfaces as nppas


sphere = vedo.shapes. IcoSphere()
features = pd.DataFrame(sphere.vertices,columns=["x","y","z"])

viewer = napari.Viewer()

layer = viewer.add_surface(nppas.to_napari_surface_data(sphere))
layer.features = features
```

Reply:
@Cryaaa , I think you were right about the contrast limits. Changing the contrast limits from [x, y] to [x, y+1] fixes the problem:

![capture](https://github.com/BiAPoL/napari-clusters-plotter/assets/38459088/38750c8b-3488-440e-bc3b-7bf3bbee378c)

Reply:
@jo-mueller Awesome, nice that it was a quick fix 🙌🏻

Reply:
Fixed in 9c7fb837078a910de39043d0f6ca112241b05946